[WIP][Bars] Option to display totals on top of each bar

wilhelmhb opened the pull request:
This PR aims at allowing users to display the total of each bar on top of it (bottom of it for negative values).

TODO:
- [X] Need description: see #85 
- [X] Base implementation
- [X] Lint
- [ ] Add tests
- [ ] Add documentation

wilhelmhb commented:
I haven't yet been able to run the tests, probably a browser/configuration issue. Thus I haven't started on the jasmine tests. I can provide a few test cases if needed.
I don't really see where the documentation is written, but if someone points me in the right direction, I'll be glad to add a few examples.

archmoj commented on src/traces/bar/layout_attributes.js:
```diff
@@ -63,5 +63,17 @@ module.exports = {
+    displaytotal: {
+        valType: 'boolean',
+        dflt: false,
+        role: 'style',
```
Please add 
```
editType: 'plot',
```
for this new attribute.

archmoj commented on src/traces/bar/layout_attributes.js:
```diff
@@ -63,5 +63,17 @@ module.exports = {
+    totaltemplate: {
+        valType: 'string',
+        dflt: '%{total}',
+        role: 'style',
```
Please add 
```
editType: 'plot',
```
for this new attribute as well.

archmoj commented on src/traces/bar/plot.js:
```diff
@@ -295,7 +298,7 @@ function appendBarText(gd, plotinfo, bar, cd, i, x0, x1, y0, y1, opts, makeOnCom
 
     var text = getText(fullLayout, cd, i, xa, ya);
-    textPosition = getTextPosition(trace, i);
+    textPosition = getTextPosition(trace, i, fullLayout.displaytotal);
```
```suggestion
    textPosition = fullLayout.barshowtotal ? 'inside' : getTextPosition(trace, i);
```

archmoj commented on src/traces/bar/plot.js:
```diff
@@ -628,7 +673,8 @@ function getText(fullLayout, cd, index, xa, ya) {
 }
 
-function getTextPosition(trace, index) {
+function getTextPosition(trace, index, forceInside) {
+    if(forceInside) return 'inside';
```
```suggestion
function getTextPosition(trace, index) {
```

archmoj commented on src/traces/bar/layout_attributes.js:
```diff
@@ -63,5 +63,17 @@ module.exports = {
         ].join(' ')
-    }
+    },
+    displaytotal: {
```
`displaytotal` is too general while this attribute lives in `layout`.
How about `barshowtotal`?

wilhelmhb replied:
Much better indeed.
Should I also rename `totaltemplate` to `bartotaltemplate`, so that each bar specific option starts with `bar` ? Or do we want to reuse `totaltemplate` for area charts as well ?

archmoj commented on src/traces/bar/layout_defaults.js:
```diff
@@ -55,4 +55,6 @@ module.exports = function(layoutIn, layoutOut, fullData) {
 
     coerce('bargap', (shouldBeGapless && !gappedAnyway) ? 0 : 0.2);
     coerce('bargroupgap');
+    coerce('displaytotal');
+    coerce('totaltemplate');
```
```suggestion
    var showTotal = coerce('barshowtotal');
    if(showTotal) coerce('totaltemplate');
```

archmoj reviewed (commented):
Thanks very much for the PR.
Before adding tests, I suggest you [sync](https://digitaldrummerj.me/git-sync-fork-to-master/) the `master` branch of fork with the latest changes of `plotly/plotly.js/master`; then `git merge upstream/master` into this branch.
Please find my other comments below:

wilhelmhb commented:
Any update on this ?

nicolaskruchten commented:
Thanks for this PR! Our team is currently working hard on finishing up the 2.0 release of Plotly.js which explains our radio-silence for the past while, and I apologize for not communicating more clearly about it. This PR should probably land in 2.1, which we expect to come out in April. I'm sorry for the long timelines here :)

vovavili commented:
@nicolaskruchten Hello! I am interested in using this feature. Now that we're reaching April 2022 and plotly 2.11, are there any updates on this PR?

archmoj commented:
@vovavili could you please fetch upstream/master and merge it into this branch?

hjc0826 commented:
I am interested in using this feature. Any update?

wilhelmhb commented:
@nicolaskruchten @archmoj I've just merged upstream/master in this branch
I removed a typo, and added an example in [test/image/mocks/bar_stackrelative_negative_total.json](https://github.com/plotly/plotly.js/pull/5375/files#diff-4144147756b000689b159ba6f841fbfbd8f329217c4eb53836d18cdb7f511f8b), available when you run `npm start`, then lookup the filename in the `mocks` search field. 

I personally have a weird result on my machine. However, as I have the same result when using any text with `textposition : outside`, I'm assuming this is an issue with plotly.js, which should be a separate matter.

I'm still unable to run the jasmine tests on my machine.

Hoping this can still make it to 2.13

hadarby commented:
Hi there is any update?

archmoj commented on src/traces/bar/layout_attributes.js:
```diff
@@ -51,5 +51,19 @@ module.exports = {
+    barshowtotal: {
+        valType: 'boolean',
+        dflt: false,
+        role: 'style',
```
There is no longer `role` needed for attributes like this one.
Please remove the `role: 'style',` line.

archmoj commented on src/traces/bar/layout_attributes.js:
```diff
@@ -51,5 +51,19 @@ module.exports = {
+    totaltemplate: {
+        valType: 'string',
+        dflt: '%{total}',
+        role: 'style',
```
There is no longer `role` needed for attributes like this one.
Please remove the role: 'style', line.

archmoj commented on src/traces/bar/layout_attributes.js:
```diff
@@ -51,5 +51,19 @@ module.exports = {
+        editType: 'plot',
+        description: 'Display the total value of each bar.'
+    },
+    totaltemplate: {
```
We may want this attribute or a similar one to control hover.
It may be better to have 
```js
stackbartotal: {
  texttemplate: ...,
  hovertemplate: ...
}
```
And there is no need for `barshowtotal` as it could be enabled by `texttemplate`.

archmoj commented:
Thanks @wilhelmhb for the great PR. 
My apologies for the late review.
Kindly find my comments regarding the new attributes.
Before adding new tests, I also suggest that you possibly pull `upstream/master` and merge it to your branch.
@alexcjohnson it would be great if you also provide a review. Thank you!

Pampelmops commented:
October 2023, I could really use this feature. Thank you.

futurejohn commented:
@archmoj @alexcjohnson How is the review going? It would be really helpful to display totals for bar charts.
